pkg/load/configAgent.go

```diff
@@ -0,0 +1,162 @@
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load configs: %v", err)
+	}
+
+	// periodic reload
+	if a.cycle != 0 {
+		interrupts.Run(func(ctx context.Context) {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(a.cycle):
+					if err := configCoalescer.Run(); err != nil {
+						log.Errorf("Failed to reload configs: %v", err)
+					}
+				}
+			}
+		})
+	}
+
+	// fsnotify reload
+	configWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create new watcher: %v", err)
+	}
+	populateWatcher(configWatcher, a.configPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to populate watcher: %v", err)
+	}
+	interrupts.Run(func(ctx context.Context) {
+		defer configWatcher.Close()
+		reloadWatcher(ctx, configWatcher, configCoalescer, a.configPath)
+	})
+	return a, nil
+}
+
+func (a *agent) GetConfig(info config.Info) (api.ReleaseBuildConfiguration, error) {
+	a.lock.Lock()
+	config, ok := a.configs[info.Basename()]
+	a.lock.Unlock()
+	if !ok {
+		return api.ReleaseBuildConfiguration{}, fmt.Errorf("Could not find config %s", info.Basename())
+	}
+	return config, nil
+}
+
+func (a *agent) GetGeneration() int {
+	return a.generation
+}
+
+// loadFilenameToConfig generates a new filenameToConfig map.
+func (a *agent) loadFilenameToConfig() error {
+	log.Debug("Reloading configs")
+	configs := filenameToConfig{}
+	err := filepath.Walk(a.configPath, func(path string, info os.FileInfo, err error) error {
+		ext := filepath.Ext(path)
+		if info != nil && !info.IsDir() && (ext == ".yml" || ext == ".yaml") {
+			configSpec, err := Config(path)
+			if err != nil {
+				return fmt.Errorf("failed to load ci-operator config (%v)", err)
+			}
+
+			if err := configSpec.ValidateAtRuntime(); err != nil {
+				return fmt.Errorf("invalid ci-operator config: %v", err)
+			}
+			log.Debugf("Adding %s to filenameToConfig", filepath.Base(path))
+			configs[filepath.Base(path)] = *configSpec
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	a.lock.Lock()
+	a.configs = configs
+	a.generation++
+	a.lock.Unlock()
+	log.Info("Configs reloaded")
+	return nil
+}
+
+func (a *agent) getConfigCoalescer() coalescer.Coalescer {
+	return coalescer.NewCoalescer(a.loadFilenameToConfig)
+}
+
+func populateWatcher(watcher *fsnotify.Watcher, root string) error {
+	return filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		// We only need to watch directories as creation, deletion, and writes
+		// for files in a directory trigger events for the directory
+		if info != nil && info.IsDir() {
+			log.Debugf("Adding %s to watch list", path)
+			err = watcher.Add(path)
+			if err != nil {
+				return fmt.Errorf("Failed to add watch on directory %s: %v", path, err)
+			}
+		}
+		return nil
+	})
+}
+
+func reloadWatcher(ctx context.Context, w *fsnotify.Watcher, c coalescer.Coalescer, path string) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
```

Comment:
Seems like you could close the watcher here for a little cleaner workflow instead of the defer above